### ISSUE TITLE
Stabilize (only) the never type

### ISSUE THREAD

**Issue by SimonSapin** [rust-lang/rust#58184]
This is an alternative to https://github.com/rust-lang/rust/issues/57012 that does not propose changing the type inference fallback (yet). This unblocks `TryFrom` and `TryInto` https://github.com/rust-lang/rust/issues/33417.

#49593 has now been fixed. This was the reason for the previous revert of stabilization (#50121).

## What is being stabilized

* `!` is now a full-fledged type and can now be used in any type position (eg. [RFC 1216](https://github.com/rust-lang/rfcs/pull/1216)). The `!` type can coerce into any other type, see https://github.com/rust-lang/rust/tree/master/src/test/run-fail/adjust_never.rs for an example.
* The `never_type` feature gate is stable, although some of the behaviours it used to gate now live behind the new `exhaustive_patterns` feature gate (see below).

## What is *not* being stabilized

* Exhaustive pattern-matching for uninhabited types. eg.

    ```rust
    let x: Result<u32, !> = ...;
    let Ok(y) = x;
    ```

  This code will still complain that `Ok(_)` is a refutable pattern. This can be fixed by using the `exhaustive_patterns` feature gate. See [RFC 1872](https://github.com/rust-lang/rfcs/pull/1872) for progress on this issue. See https://github.com/rust-lang/rust/tree/master/src/test/ui/feature-gate-exhaustive-patterns.rs for the testcase which confirms that this behaviour is still gated.

* The type inference changes proposed in https://github.com/rust-lang/rust/issues/48950 are not part of this proposal, since they can break some previously-valid programs and the mitigations discussed in https://github.com/rust-lang/rust/issues/57012 are not ready yet.

  This change is moved to a new dedicated feature gate, `inference_fallback_to_never`, instead of `never_type`.

**Comment by withoutboats** [comment]
@rfcbot fcp merge

There is currently a lack of consensus regarding the change to fallback and how to handle edge case breakages. I don't believe there is any reason this change has to be tied together with making `!` usable from stable, for which we have a strong consensus and a lot of groundwork laid. I think its a reasonable way to move forward for us to separate these two changes, so we can stabilize `!` now without blocking it (and by extension, `TryFrom`) on reaching consensus about the fallback changes.

**Comment by SimonSapin** [comment]
It was pointed out to be that `#![feature(never_type)]` changes inference in current Nightly. I’ve edited the proposal to mention moving that to a new feature gate.

**Comment by cramertj** [comment]
@rfcbot concern motivation-to-change-fallback

I'm concerned that if we stabilize `!` now without the fallback change then there won't be sufficient motivation for us to land the fallback change later, and we'll lose out on a feature that would have made the language better. IMO we should go ahead with the full stabilization-- there are inference changes that we don't currently warn against, but we've traditionally allowed such breakages in other places and I don't believe this change is more significant or detrimental.

**Comment by Centril** [comment]
@rfcbot concern eddybs-concern-on-the-other-issue

See https://github.com/rust-lang/rust/issues/57012#issuecomment-460740678.

**Comment by mati865** [comment]
What is the status on this?
There was no voting activity since 18 April.

**Comment by SimonSapin** [comment]
As far as I know the status has not significantly changed since my summary it https://github.com/rust-lang/rust/issues/57012#issuecomment-467889706.

**Comment by withoutboats** [comment]
The concern labelled "eddybs-concern-on-the-other-issue" has been resolved by #62661.

**Comment by Centril** [comment]
Indeed it has.

@rfcbot resolve eddybs-concern-on-the-other-issue

**Comment by cramertj** [comment]
I'll resolve my concern on this (motivation-to-change-fallback) pending a lang team discussion as proposed in https://github.com/rust-lang/rust/pull/65355#issuecomment-541818923.

**Comment by cramertj** [comment]
Sorry for the labels nonsense-- something is going wacky with GH in my browser.
